# Introduction

This user car dataset is scrapped from the German eBay site. Originally uploaded to Kaggle, this version is a dirtied version with 50,000 datapoints.

Aim of this project is to clean the data and analyze the listings.

In [1]:
import pandas as pd
import numpy as np

In [2]:
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

* 5 columns have missing values (vehicleType, gearbox, model, fuelType, notRepairedDamage)
* Date fields need to be converted to datetime
* Price field needs to be converted int or float (a.k.a strip out the dollar sign)
* Odometer field similar needs to strip out the 'km'.
* Column names follow the camelcase name convension.

### Convert Column Names
* convert from camelcase to snakecase

In [5]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
autos_columns = autos.columns.copy()

In [7]:
col_mapping = {'yearOfRegistration' : 'registration_year', 
               'monthOfRegistration' : 'registration_month',
               'notRepairedDamage' : 'unrepaired_damage', 
               'dateCreated' : 'ad_created',
               'dateCrawled' : 'date_crawled', 
               'offerType' : 'offer_type', 
               'vehicleType' : 'vehicle_type',
               'powerPS' : 'power_ps', 
               'fuelType' : 'fuel_type', 
               'nrOfPictures' : 'nr_of_pictures',
               'postalCode' : 'postal_code',
               'abtest' : 'ab_test',
               'lastSeen' : 'last_seen'}

In [8]:
new_col = []
for col in autos.columns:
    if col in col_mapping:
        col = col_mapping[col]
    else:
        col = col
    new_col.append(col)

In [9]:
autos.columns = new_col

In [10]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

#### Reasons for column name changes:
* adhering to Python column naming convention, convert from camelcase to snakecase
* yearOfRegistration, monthOfRegistration, notRepairedDamage, dateCreated were renamed to be more clear and data relevant. For example, dateCreated is changed to ad_created.

### Data Exploration
* what else needs to be cleaned?
    * columns where all or almost all values are the same
    * numeric data stored as text
    * date columns stored as text
    * etc.

In [11]:
autos.describe(include = 'all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [12]:
autos['nr_of_pictures'].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

In [13]:
autos['odometer'].value_counts()

150,000km    32424
125,000km     5170
100,000km     2169
90,000km      1757
80,000km      1436
70,000km      1230
60,000km      1164
50,000km      1027
5,000km        967
40,000km       819
30,000km       789
20,000km       784
10,000km       264
Name: odometer, dtype: int64

In [14]:
autos['price'].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

#### Findings

The 'nr_of_pictures' column can be dropped since it only has a single value. This won't be valuable in any analysis.

Two columns ('seller' and 'offer_type') have mostly singular value. 

The 'odometer' column needs to be cleaned. It's a text field that needs to be converted to numerical (most likely int).

The 'price' column also needs to be cleaned. Need to strip out the dollar sign and convert to int.

The 'registration_year' minimum and maximum values are very unrealistic. Don't believe we can have cars from the year 1000 or from year 9999.

Similarly, 'power_ps' has a minimum value of 0 and max of 17700. The 0 value should probably be NaN and the higher values should be cleaned up.

In [15]:
# clean 'odometer' column and convert to numerical
autos['odometer'] = autos['odometer'].str.replace(',', '')
autos['odometer'] = autos['odometer'].str.replace('km', '')
autos['odometer'] = autos['odometer'].astype(int)

In [16]:
autos.rename(columns = {'odometer' : 'odometer_km'}, inplace = True)

In [17]:
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [18]:
# clean 'price' column and convert to numerical
autos['price'] = autos['price'].str.replace(',', '')
autos['price'] = autos['price'].str.replace('$', '')
autos['price'] = autos['price'].astype(int)

In [19]:
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

### Removing Outliers

In [20]:
# how many unique values
autos['odometer_km'].unique().shape

(13,)

In [21]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [22]:
autos['odometer_km'].value_counts().sort_index(ascending = False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

There are 13 unique odometer values with the values tend to trend on the higher side. There doesn't seem to be outliers that pop out.

In [23]:
# number of unique prices
autos['price'].unique().shape

(2357,)

In [24]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

With the max price at \\$100,000,000, there are some outliers to clean up. Conversely, the \\$0 price tag looks reasonable. Used car owners could just want to get rid of the cars.

In [25]:
autos['price'].value_counts().sort_index(ascending = True).head(10)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64

Indeed, quite a spike at the \$0.

In [26]:
autos['price'].value_counts().sort_index(ascending = False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

These prices **may** be reasonable if these cars are high end. Let's take a deeper look at the car models.

In [27]:
autos[autos['price'] > 100000][['brand', 'price']].sort_values('price')

brand     price
29286         porsche    104900
17540         porsche    105000
16964  sonstige_autos    105000
49391  sonstige_autos    109999
22060  sonstige_autos    114400
7402          porsche    115000
21783         porsche    115991
33884         porsche    116000
38814         porsche    119500
43282         porsche    119900
2751          porsche    120000
44406   mercedes_benz    120000
8232          porsche    128000
1878          porsche    129000
49815            ford    130000
14268         porsche    135000
2454          porsche    137999
32185         porsche    139997
49668         porsche    145000
33638         porsche    151990
10500         porsche    155000
45387         porsche    163500
22673         porsche    163991
18509         porsche    169000
32840   mercedes_benz    169999
11433            audi    175000
17140   mercedes_benz    180000
20351             bmw    190000
28090  sonstige_autos    194000
43668         porsche    197000
40918         porsche    198000
37840         porsche    220000
38299             bmw    250000
47337             bmw    259000
12682         porsche    265000
35923         porsche    295000
34723         porsche    299000
14715  sonstige_autos    345000
36818         porsche    350000
37585      volkswagen    999990
43049      volkswagen    999999
514              ford    999999
22947             bmw   1234566
7814   sonstige_autos   1300000
47634  sonstige_autos   3890000
11137  sonstige_autos  10000000
2897             ford  11111111
24384      volkswagen  11111111
27371            fiat  12345678
47598            opel  12345678
39377           volvo  12345678
42221         citroen  27322222
39705   mercedes_benz  99999999

There are some ridiculous entries like a Ford for \\$999,999 and a Volkswagen for \\$11,111,111 to name a couple. Looks like \\$350,000 is a good cut off point. There a big jump from \\$350,000 to \\$999,999 and the nonsensical data points also show up when prices are greater than \\$999,999.

In [28]:
autos = autos[autos['price'] <= 350000]

In [29]:
autos['price'].describe()

count     49986.000000
mean       5721.525167
std        8983.617820
min           0.000000
25%        1100.000000
50%        2950.000000
75%        7200.000000
max      350000.000000
Name: price, dtype: float64

This removed 24 high price entries.

### Explore the date columns

There are 3 columns where the dates are represented as strings. We'll extract the date part of the string from date_crawled, ad_created, and last_seen.

In [30]:
autos['date_crawled'].str[:10].value_counts().sort_index()

2016-03-05    1269
2016-03-06     697
2016-03-07    1798
2016-03-08    1663
2016-03-09    1660
2016-03-10    1606
2016-03-11    1624
2016-03-12    1838
2016-03-13     778
2016-03-14    1831
2016-03-15    1699
2016-03-16    1475
2016-03-17    1575
2016-03-18     653
2016-03-19    1745
2016-03-20    1891
2016-03-21    1874
2016-03-22    1645
2016-03-23    1619
2016-03-24    1455
2016-03-25    1587
2016-03-26    1624
2016-03-27    1552
2016-03-28    1742
2016-03-29    1707
2016-03-30    1681
2016-03-31    1595
2016-04-01    1690
2016-04-02    1770
2016-04-03    1934
2016-04-04    1824
2016-04-05     655
2016-04-06     159
2016-04-07      71
Name: date_crawled, dtype: int64

The date_crawled column contains dates from 3/5/2016 to 4/7/2016.

In [31]:
autos['ad_created'].str[:10].value_counts().sort_index()

2015-06-11       1
2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
              ... 
2016-04-03    1946
2016-04-04    1842
2016-04-05     592
2016-04-06     163
2016-04-07      64
Name: ad_created, Length: 76, dtype: int64

Seems most of the ads were scrapped the day they were created.

In [32]:
autos['last_seen'].str[:10].value_counts().sort_index()

2016-03-05       54
2016-03-06      221
2016-03-07      268
2016-03-08      379
2016-03-09      492
2016-03-10      538
2016-03-11      626
2016-03-12     1190
2016-03-13      449
2016-03-14      640
2016-03-15      794
2016-03-16      822
2016-03-17     1396
2016-03-18      371
2016-03-19      787
2016-03-20     1035
2016-03-21     1036
2016-03-22     1079
2016-03-23      929
2016-03-24      978
2016-03-25      960
2016-03-26      848
2016-03-27      801
2016-03-28     1042
2016-03-29     1116
2016-03-30     1242
2016-03-31     1191
2016-04-01     1155
2016-04-02     1244
2016-04-03     1268
2016-04-04     1231
2016-04-05     6212
2016-04-06    11046
2016-04-07     6546
Name: last_seen, dtype: int64

The distribution of counts seem to increase the more recent the dates get. Possible explanation is older listings are either taken down due to transaction going through or due to inactivity.

#### And let's take a look at registration_year

In [33]:
autos['registration_year'].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Seems the average registration year is 2005. It's difficult to  assess percentile accuracy when there are years from a 1000 years ago to 6000 years into the future.

### Dealing with registration_year

Cars registered after the listing was seen is impossible, so any year above 2016 is incorrect. But how about the earliest valid year? Let's explore a bit more.

In [34]:
autos[~autos['registration_year'].between(1900, 2016)]['registration_year'].value_counts().sort_index()

1000       1
1001       1
1111       1
1500       1
1800       2
2017    1452
2018     491
2019       3
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       2
9996       1
9999       4
Name: registration_year, dtype: int64

Looks like we can remove years before 1900 and after 2016, BUT 2017 and 2018 have extremely high numbers. Removing these two years can skew the data, so let's keep them in for now. Perhaps these two years were user errors.

In [35]:
autos = autos[autos['registration_year'].between(1900, 2018)]

In [37]:
autos['registration_year'].max()

2018

In [39]:
autos['registration_year'].value_counts(normalize = True).sort_index()[40:]

1979    0.000701
1980    0.001942
1981    0.000600
1982    0.000861
1983    0.001061
1984    0.001061
1985    0.002082
1986    0.001521
1987    0.001501
1988    0.002842
1989    0.003623
1990    0.007906
1991    0.007126
1992    0.007806
1993    0.008907
1994    0.013211
1995    0.026262
1996    0.028904
1997    0.040593
1998    0.049100
1999    0.060009
2000    0.067135
2001    0.054084
2002    0.050702
2003    0.054585
2004    0.054785
2005    0.060349
2006    0.054184
2007    0.046118
2008    0.044657
2009    0.041974
2010    0.031966
2011    0.032707
2012    0.026482
2013    0.016133
2014    0.013311
2015    0.007987
2016    0.026342
2017    0.029064
2018    0.009828
Name: registration_year, dtype: float64

Looks like most of the cars were registered roughly between 1994 to 2014.

### Exploring Car Brands

In [40]:
autos['brand'].value_counts()

volkswagen        10679
opel               5457
bmw                5427
mercedes_benz      4731
audi               4283
ford               3477
renault            2404
peugeot            1456
fiat               1307
seat                940
skoda               785
mazda               757
nissan              754
smart               701
citroen             699
toyota              617
sonstige_autos      538
hyundai             488
volvo               456
mini                424
mitsubishi          404
honda               398
kia                 356
alfa_romeo          328
porsche             294
suzuki              293
chevrolet           283
chrysler            181
dacia               129
daihatsu            128
jeep                109
subaru              108
land_rover           99
daewoo               79
saab                 79
trabant              77
jaguar               77
rover                69
lancia               57
lada                 31
Name: brand, dtype: int64

Many of the top brands are German with Volkswagen being the most popular by far. We'll include brands with counts above the lower bound of 2%, or 1000 cars.

In [47]:
brand_average_price = {}

brand_count_normalized = autos['brand'].value_counts(normalize = True)
brand_list = brand_count_normalized[brand_count_normalized > 0.02].index

for brand in brand_list:
    brand_average_price[brand] = autos[autos['brand'] == brand]['price'].mean()

In [48]:
brand_average_price

{'volkswagen': 5159.401629366045,
 'opel': 2842.8246289169874,
 'bmw': 8028.474479454579,
 'mercedes_benz': 8380.637920101459,
 'audi': 8965.560354891431,
 'ford': 3626.5429968363533,
 'renault': 2351.301996672213,
 'peugeot': 3010.8688186813188,
 'fiat': 2697.6771231828616}

Appears to be 3 tiers among the top brands:
1. BMW, Mercedez Benz, and Audi average above \\$8,000 range
2. Volkwagen average about \\$5,000
3. Opel, Ford, Renault, Peugeot, and Fiat post around the \\$3,000 range

### Exploring link between average mileage and mean price

Next, we'll explore relationships (if any) between the mean mileage and mean price for each of the top brands.

Let's start by creating a dictionary of each brand and its respective average mileage.

In [50]:
avg_mileage = {}

for brand in brand_list:
    avg_mileage[brand] = autos[autos['brand'] == brand]['odometer_km'].mean()

Now let's create a dictionary for each brand and it's respective average price.

In [53]:
avg_price = {}

for brand in brand_list:
    avg_price[brand] = autos[autos['brand'] == brand]['price'].mean()

Finally, create a dataframe with the mileage and price as columns. We'll start with creating Series objects for each dictionary.

In [54]:
mileage = pd.Series(avg_mileage)
price = pd.Series(avg_price)

In [57]:
df = pd.DataFrame(mileage, columns = ['avg_mileage'])

In [58]:
df['avg_price'] = price

In [63]:
df.sort_values('avg_price')

avg_mileage    avg_price
renault        128223.793677  2351.301997
fiat           117012.241775  2697.677123
opel           129361.370717  2842.824629
peugeot        127352.335165  3010.868819
ford           124153.005464  3626.542997
volkswagen     129006.461279  5159.401629
bmw            132540.998710  8028.474479
mercedes_benz  130933.206510  8380.637920
audi           129643.941163  8965.560355

The average mileage seem to hover at around 130,000 km among the top brands. The luxury brands (Audi, Mercedez Benz, BMW) hold their value very well dispite having  some of the highest average mileage.

### Possible next steps

#### Data Cleaning:
1. Translate categorical data that uses german words into their english translations.
2. Convert dates to be uniform numeric data, a.k.a '2016-03-21' becomes 20160321.
3. Are there particular keywords in the name columns that can be extracted as new columns?

#### Analysis:
1. What are the most common brand/model combinations?
2. By splitting 'odometer_km' into groups, can we see if there are patterns between average prices and average mileage?
3. What's the price difference between cars with damage and cars with no damage?